docs/ldml/tr35-keyboards.md

```diff
@@ -439,8 +439,202 @@ For example, the sequence `e\m{marker}\u{300}` would be output in NFC as `è`. T
 
 ### Normalization-safe Segments
 
-For purposes of the above discussion, "normalization-safe segments" are defined as a string of codepoints which are (1) already in [NFD](https://www.unicode.org/reports/tr15/#Norm_Forms), and (2) begin with a character with [Canonical Combining Class](https://www.unicode.org/reports/tr44/#Canonical_Combining_Class_Values) of `0`. See [UAX #15 Section 9.1: Stable Code Points](https://www.unicode.org/reports/tr15/#Stable_Code_Points) for related discussion.  Text under consideration can be segmented by locating such characters.
-
+For purposes of the above discussion, "normalization-safe segments" are defined as a string of codepoints which are
+(1) already in [NFD](https://www.unicode.org/reports/tr15/#Norm_Forms), and
+(2) begin with a character with [Canonical Combining Class](https://www.unicode.org/reports/tr44/#Canonical_Combining_Class_Values) of `0`.
+See [UAX #15 Section 9.1: Stable Code Points](https://www.unicode.org/reports/tr15/#Stable_Code_Points) for related discussion.
+Text under consideration can be segmented by locating such characters.
+
+### Pseudocode for Marker-safe Normalization
+
+The following is an informative section with some example pseudocode for marker-safe normalization into NFD.
+
+1. Markers are not text. This pseudocode assumes an implementation which is able to distinguish markers from plain text.
+
+2. Some optimization notes are given, but the emphasis here is on a clear and correct implementation.
+
+3. A distinction would need to be made between (A) markers in text, such as a keystroke emitting `\m{someMarker}` or a marker in input context, and (B) markers in transform match strings, such as `<transform from="\m{marker}">` or `<transform from="\m{.}">`.  An implementation could pass a parameter through the following code path distinguishing A and B, this distinction is not shown, for clarity.
+
+Example use:
+
+- `nfdMarkers('e\u{0300}' + Marker('marker1') + '\u{0320}a\u{0300}' + Marker('marker2') + '\u{0320}')`
+- Returns: `'e' + Marker('marker1') + '\u{0320}\u{0300}a' + Marker('marker2') + '\u{0320}\u{0300}'`
```

Comment:
It may be worth splitting the input and output strings to make the "normalization-safe segment" boundaries easier to parse.  Since you use `'\u{0300}'` and `'\u{0320}'` following both the `e` and the `a`, it's easy to confuse what's happening in the center part.

On first glance, it looked to me like `'\u{0320}a\u{0300}'` had the `'a'` moved forward within a segment.  After pausing and thinking through it, I know that's not actually what's happening... but I had to pause and think through it, and I've been part of related discussions.  Fresh readers may struggle more without a bit of segment-parsing assistance.

Using different `\u` codepoints after the `'a'` would likely help as well.

```suggestion
- `nfdMarkers('e\u{0300}' + Marker('marker1') + '\u{0320}' + 'a\u{0300}' + Marker('marker2') + '\u{0320}')`
- Returns: `'e' + Marker('marker1') + '\u{0320}\u{0300}' + 'a' + Marker('marker2') + '\u{0320}\u{0300}'`
```

Reply:
> It may be worth splitting the input and output strings to make the "normalization-safe segment" boundaries easier to parse.

But the point is that they aren't split going in, and this function splits them for processing.

This is "example 3" above in the spec text. The purpose of this content here is to show the function call usage.  How about if I just say that this is "example 3" and make sure that the spec is clear about how things work?  I don't think the spec should be repeated in the pseudocode.

> Using different \u codepoints after the 'a' would likely help as well.

That would not show the need for the function.  If the glue characters were distinct, there would be no need for segmentation- you could run the `nfdMarkersSegment()` function on the entire string.

Reply:
> Since you use '\u{0300}' and '\u{0320}' following both the e and the a, it's easy to confuse what's happening in the center part.

I agree with the comment. Don't use duplicate characters in the example when the positions may change.